packages/database/textdb/textkv.go

```diff
@@ -0,0 +1,291 @@
+	return &textKV{
+		filename:      s.filename,
+		log:           s.log,
+		realm:         realm,
+		inMemoryStore: s.inMemoryStore,
+	}
+}
+
+// Realm returns the configured realm.
+func (s *textKV) Realm() kvstore.Realm {
+	return byteutils.ConcatBytes(s.realm)
+}
+
+// Shutdown marks the store as shutdown.
+func (s *textKV) Shutdown() {
+}
+
+func (s *textKV) load() (map[string]interface{}, error) {
+	data, err := os.ReadFile(s.filename)
+	if err != nil {
+		return nil, err
+	}
+	ret := map[string]interface{}{}
+	err = s.unmarshal(data, &ret)
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
+}
+
+// Iterate iterates over all keys and values with the provided prefix. You can pass kvstore.EmptyPrefix to iterate over all keys and values.
+func (s *textKV) Iterate(prefix kvstore.KeyPrefix, kvConsumerFunc kvstore.IteratorKeyValueConsumerFunc) error {
+	s.RLock()
+	defer s.RUnlock()
+	return s.inMemoryStore.Iterate(byteutils.ConcatBytes(s.realm, prefix), kvConsumerFunc)
+}
+
+// IterateKeys iterates over all keys with the provided prefix. You can pass kvstore.EmptyPrefix to iterate over all keys.
+func (s *textKV) IterateKeys(prefix kvstore.KeyPrefix, consumerFunc kvstore.IteratorKeyConsumerFunc) error {
+	s.RLock()
+	defer s.RUnlock()
+	return s.inMemoryStore.IterateKeys(byteutils.ConcatBytes(s.realm, prefix), consumerFunc)
+}
+
+// clear the key/value store
+func (s *textKV) Clear() error {
+	s.Lock()
+	defer s.Unlock()
+	err := s.inMemoryStore.Clear()
+	if err != nil {
+		return err
+	}
+	return os.WriteFile(s.filename, []byte("{}"), storePerm)
+}
+
+// Get gets the given key or nil if it doesn't exist or an error if an error occurred.
+func (s *textKV) Get(key kvstore.Key) (value kvstore.Value, err error) {
+	s.RLock()
+	defer s.RUnlock()
+	return s.inMemoryStore.Get(byteutils.ConcatBytes(s.realm, key))
+}
+
+// Set sets the given key and value.
+func (s *textKV) Set(key kvstore.Key, value kvstore.Value) error {
+	s.Lock()
+	defer s.Unlock()
+
+	// first update in inMemoryStore
+	err := s.inMemoryStore.Set(byteutils.ConcatBytes(s.realm, key), value)
+	if err != nil {
+		return err
+	}
+
+	return s.flush()
+}
+
+// Has checks whether the given key exists.
+func (s *textKV) Has(key kvstore.Key) (bool, error) {
+	s.RLock()
+	defer s.RUnlock()
+	return s.inMemoryStore.Has(byteutils.ConcatBytes(s.realm, key))
+}
+
+// Delete deletes the entry for the given key.
+func (s *textKV) Delete(key kvstore.Key) error {
+	s.Lock()
+	defer s.Unlock()
+
+	err := s.inMemoryStore.Delete(byteutils.ConcatBytes(s.realm, key))
+	if err != nil {
+		return err
+	}
+
+	return s.flush()
+}
+
+// DeletePrefix deletes all the entries matching the given key prefix.
+func (s *textKV) DeletePrefix(prefix kvstore.KeyPrefix) error {
+	s.Lock()
+	defer s.Unlock()
+
+	err := s.inMemoryStore.DeletePrefix(byteutils.ConcatBytes(s.realm, prefix))
+	if err != nil {
+		return err
+	}
+	return s.flush()
+}
+
+func (s *textKV) flush() error {
+	var err error
```

Comment:
we probably also want to lock on flush
```suggestion
	s.Lock()
	defer s.Unlock()
	var err error
```

Reply:
`flush` is only called within code that already has a handle to the lock. So calling s.Lock() inside flush will may block as the parent code still has not unlocked. Or did I think of this wrongly.

Reply:
oh I see, I thought it was a public method, now that I see the public Flush just returns nil. nevermind :+1: